File: backend/api/_utils/auth_middleware.py
"""
Authentication middleware for VrikshAI.

Provides JWT token generation and verification for secure API access.
"""

import os
import jwt
import logging
from datetime import datetime, timedelta
from typing import Dict, Any
from http.server import BaseHTTPRequestHandler

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Token expiration: 7 days
TOKEN_EXPIRATION_DAYS = 7


def generate_token(user_id: str, email: str) -> str:
    """
    Generate a JWT token for authenticated user.

    Args:
        user_id: User's UUID
        email: User's email address

    Returns:
        JWT token string

    Raises:
        ValueError: If JWT_SECRET is not set
    """
    secret = os.getenv('JWT_SECRET')
    if not secret:
        raise ValueError("JWT_SECRET environment variable is required")

    payload = {
        'user_id': user_id,
        'email': email,
        'exp': datetime.utcnow() + timedelta(days=TOKEN_EXPIRATION_DAYS),
        'iat': datetime.utcnow()
    }

    token = jwt.encode(payload, secret, algorithm='HS256')
    logger.info(f"Token generated for user {user_id}")
    return token


def verify_token(request: BaseHTTPRequestHandler) -> Dict[str, Any]:
    """
    Verify JWT token from request Authorization header.

    Args:
        request: HTTP request handler with headers

    Returns:
        Dictionary with user_id and email

    Raises:
        Exception: If token is missing, expired, or invalid
    """
    try:
        # Extract Authorization header
        auth_header = request.headers.get('Authorization')
        if not auth_header:
            raise Exception("Missing Authorization header")

        # Extract Bearer token
        parts = auth_header.split()
        if len(parts) != 2 or parts[0].lower() != 'bearer':
            raise Exception("Invalid Authorization header format. Use: Bearer <token>")

        token = parts[1]

        # Verify token
        secret = os.getenv('JWT_SECRET')
        if not secret:
            raise ValueError("JWT_SECRET environment variable is required")

        payload = jwt.decode(token, secret, algorithms=['HS256'])

        logger.info(f"Token verified for user {payload['user_id']}")
        return {
            'user_id': payload['user_id'],
            'email': payload['email']
        }

    except jwt.ExpiredSignatureError:
        logger.warning("Expired token attempt")
        raise Exception("Token has expired. Please login again.")

    except jwt.InvalidTokenError as e:
        logger.warning(f"Invalid token attempt: {str(e)}")
        raise Exception("Invalid token. Please login again.")

    except Exception as e:
        logger.error(f"Token verification failed: {str(e)}")
        raise

File: api/_utils/auth_middleware.js
/**
 * Authentication middleware for VrikshAI.
 *
 * Provides JWT token generation and verification for secure API access.
 */

import jwt from 'jsonwebtoken';

// Token expiration: 7 days
const TOKEN_EXPIRATION_DAYS = 7;

/**
 * Generate a JWT token for authenticated user.
 *
 * @param {string} userId - User's UUID
 * @param {string} email - User's email address
 * @returns {string} JWT token string
 * @throws {Error} If JWT_SECRET is not set
 */
export function generateToken(userId, email) {
  const secret = process.env.JWT_SECRET;
  if (!secret) {
    throw new Error('JWT_SECRET environment variable is required');
  }

  const now = Math.floor(Date.now() / 1000);
  const expirationTime = now + (TOKEN_EXPIRATION_DAYS * 24 * 60 * 60);

  const payload = {
    user_id: userId,
    email: email,
    exp: expirationTime,
    iat: now
  };

  const token = jwt.sign(payload, secret, { algorithm: 'HS256' });
  console.log(`Token generated for user ${userId}`);
  return token;
}

/**
 * Verify JWT token from request Authorization header.
 *
 * @param {Object} req - HTTP request object with headers
 * @returns {Object} Object with user_id and email
 * @throws {Error} If token is missing, expired, or invalid
 */
export function verifyToken(req) {
  try {
    // Extract Authorization header
    const authHeader = req.headers.authorization || req.headers.Authorization;
    if (!authHeader) {
      throw new Error('Missing Authorization header');
    }

    // Extract Bearer token
    const parts = authHeader.split(' ');
    if (parts.length !== 2 || parts[0].toLowerCase() !== 'bearer') {
      throw new Error('Invalid Authorization header format. Use: Bearer <token>');
    }

    const token = parts[1];

    // Verify token
    const secret = process.env.JWT_SECRET;
    if (!secret) {
      throw new Error('JWT_SECRET environment variable is required');
    }

    const payload = jwt.verify(token, secret, { algorithms: ['HS256'] });

    console.log(`Token verified for user ${payload.user_id}`);
    return {
      user_id: payload.user_id,
      email: payload.email
    };
  } catch (error) {
    if (error.name === 'TokenExpiredError') {
      console.error('Expired token attempt');
      throw new Error('Token has expired. Please login again.');
    } else if (error.name === 'JsonWebTokenError') {
      console.error('Invalid token attempt:', error.message);
      throw new Error('Invalid token. Please login again.');
    } else {
      console.error('Token verification failed:', error.message);
      throw error;
    }
  }
}
